stream.go

```diff
@@ -190,14 +191,36 @@ func (s *stream) Write(p []byte) (int, error) {
 	s.onData()
 	s.mutex.Unlock()
 
+	var err error
 	for {
 		s.mutex.Lock()
+		deadline := s.writeDeadline
+		s.mutex.Unlock()
+
+		if !deadline.IsZero() && !time.Now().Before(deadline) {
+			err = errDeadline
+			break
+		}
+
+		s.mutex.Lock()
 		if s.dataForWriting == nil || s.err != nil {
 			s.mutex.Unlock()
 			break
 		}
 		s.mutex.Unlock()
-		<-s.writeChan
+
+		if deadline.IsZero() {
+			<-s.writeChan
+		} else {
+			select {
+			case <-s.writeChan:
+			case <-time.After(deadline.Sub(time.Now())):
```

Comment:
I don't really see a race condition here, just an optimization to save one syscall.
In the situation you're describing the timer immediately fires, so the loop will loop and we'll break. If I replace this by one call of `time.Now()`, the timer will fire a fraction of a millisecond later, and the loop will loop and break.